### ISSUE TITLE
Sparrow Server disconnect with mix coordinator 502 error and StompSession exception

### ISSUE THREAD

**Issue by s373nZ** [sparrowwallet/sparrow#1097]
I'm running Sparrow Server 1.7.9 on a Raspibolt stack on RPi with 8GB memory. Today I checked to make sure Sparrow was still mixing and found it in a disconnected state. I've reviewed the logs and found some error reports related to 502 errors from Samourai coordinator `pool.whirl.mx`. Logs are attached. Logs end there, so I feel my process went into a disconnected state after this. Is it possible that Sparrow Server did not recover gracefully from the 502 error and caused the disconnect?

I've also reviewed my Fulcrum logs and saw that it lost connection to bitcoind for about 15 mins around that time, and bitcoin logs don't show an outage or error, but suggest it detected a potential stale tip around that time.

I'm posting here in hopes that Sparrow can gracefully recover from events like this. Let me know if you need more information or logs -- happy to help where I can.

[sparrow-2023-09-12.log](https://github.com/sparrowwallet/sparrow/files/12594871/sparrow-2023-09-12.log)

**Comment by craigraw** [comment]
Connection errors to the Whirlpool coordinator will not cause Sparrow to disconnect from Fulcrum, so these events are unrelated. 

With respect to Fulcrum and bitcoind, neither perform logging to indicate client disconnections with default configuration. Sparrow also doesn't log this normally, as it would create an excessively large log file. 

It may be that Fulcrum experienced some difficulty that caused it to disconnect Sparrow, but of course it's hard to say with available information. Under these circumstances Sparrow will try to reconnect after every minute (initially), backing off gradually in frequency over time. If you turn on `DEBUG` logging in Sparrow (see FAQ), you can see the connection attempts.

You will likely experience a more reliable connection with better server hardware, but that's another discussion :)

**Comment by RequestPrivacy** [comment]
Your problem sounds like a similiar thing I had. See my answer [here](https://github.com/sparrowwallet/sparrow/issues/899#issuecomment-1636324036).

I guess it has to do with hardware limitations of the pi in combination with an increased address count and/or a high gap limit - try to mix out of your wallet to a fresh one (be sure to mix to the postmix account of the new wallet so that you can continue to remix there) and see if it solved the problem. It sure did for me...

**Comment by s373nZ** [comment]
@RequestPrivacy Thanks for chiming in. I am definitely experiencing similar behavior. I've been confused as to why I can reconnect immediately after just quitting and restarting Sparrow Server. You think this was a result of deep mixing in the multiple thousands of addresses? The postmix wallet in question here for me has under 600 transactions and the target "mix out" wallet has under 100.

Did you also upgrade your hardware? Please do stay in touch on this if you notice anything.

@craigraw Yes, better hardware would be the easy answer here, but I'm sort of sticking to the Raspibolt stack to practice diagnosing issues in a constrained resource environment... and masochism. Thanks for your patience and support on my Github issues around this over the last few months. I'll be running with debug log enabled to see if I can get a better view of the reconnect attempts.

**Comment by RequestPrivacy** [comment]
I mixed out to a new wallet and the first couple of days all worked like a charm (before the mix out I had to reconnect manually almost every day).

Then the disconnected states happened again although sparrow only used roughly 200 addresses up to this point. In the settings of the postmix account (under "Advanced) I saw that in a matter of 1 or 2 days Sparrow Server automatically increased to gap limit to nearly 3000, which is insane in that short amount of time imho. I don't know how the gap limit is important during polling fulcrum though...
-> So maybe it's more about the gap limit and not the "deepness" of the wallet?


Another observation I made is that decreasing the amount of utxos in the postmix account also seems to help. I mean, after a certain point of mixing out, the "old/bad" wallet started to mix properly again i.e. it didn't disconnect so often anymore.
-> Maybe it's a combination of number of utxos in combination with high gap limit? 

Do you have many utxos in the wallet? And can you tell us the gap limit sparrow set?

**Comment by s373nZ** [comment]
There are currently 178 UTXOs in the postmix wallet, and sparrow has set a gap limit of 2369 for that wallet, which sounds close to what you are reporting. Do you recommend manually changing the gap limit?

Another thing I am trying now is connecting via pure TCP instead of SSL. I feel kind of dumb for not trying this sooner, but I opted for it in a conservative moment and sort of forgot about it. We'll see.

**Comment by RequestPrivacy** [comment]
> Do you recommend manually changing the gap limit?

No, I don't think that would be helpful.

It would be interesting to see if your connection improves like mine after you reduced the utxo set down to only a couple of utxos (e.g. from your 178 down to maybe 20 bigger ones).
 
I understand this is much to ask since you might not be comfortable to consolidate so drastically and it potentially incurs new mixing fees and non-trivial planning during consolidation and whether there is an actionable result at the end is questionable.

Maybe better: keep mixing out to the new wallet but don't mix with the new one (only mix with the wallet which mixes out). After some time almost all of your 178 utxos are in the new wallet and if we have the same issue the connection drops will markedly reduce after some threshold.

> Another thing I am trying now is connecting via pure TCP instead of SSL.

Already done that, didn't improve it for me. But as a first step, go for it!

**Comment by s373nZ** [comment]
@RequestPrivacy I've been doing some analysis with system monitoring tools including `iotop`, `iostat` along with some others. I managed to run some tests during a time when the box seemed to responding sluggishly. The interesting thing was that `iostat`, after a while, eventually reported some consistently high `%iowait` times, in the > 70% range. My currently theory is that the drive access over USB connection is causing the issue and I/O wait times are holding up the rest of the system under certain circumstances.

I've been down the rabbit hole a few years ago with the PI storage over USB as many people encounter issues with interfaces that don't support UAS. That got me to buy the Geekworm setup with the X825 board, which I believed has helped and my system, and reports the driver using UAS. I'm guessing, though, that this I/O limitation is still having a limited impact when stacking too many services on the Pi. I believe that the Samourai team has dropped "official support" for the RPi due to hardware limitations, and they've issued a warning about the external disk / UAS complications to custom build users for a while, while standardizing on integrated disk access over PCI and RockPro64 for their product.

Still learning a lot about reading these diagnostics and practicing to apply the tools correctly, but thought I'd share my current hypothesis.

In the meantime, I've got a refurbished minipc on order and will be working to migrate my node to that in the coming weeks.
